src/wallet/wallet2.cpp

```diff
@@ -11069,7 +11069,8 @@ std::string wallet2::export_outputs_to_str() const
 
   std::stringstream oss;
   boost::archive::portable_binary_oarchive ar(oss);
-  ar << export_outputs();
+  const auto& outputs = export_outputs();
```

Comment:
No &, or the following might read a deleted object.

Reply:
No

Reply:
Why not ?

Reply:
Trust me, i know how the C++ works.

Reply:
While a simple test case appears to support your opinion, "trust me" just isn't going to work :)
Can you explain why this is not accessing a ref to unwound stack space, since the object isn't copied to something in scope ?

Reply:
"Whenever a reference is bound to a temporary or to a subobject thereof, the lifetime of the temporary is extended to match the lifetime of the reference, [with the following exceptions](https://en.cppreference.com/w/cpp/language/reference_initialization#Lifetime_of_a_temporary): "

Reply:
I would say that & is only confusing and not needed here anyway. export_outputs is defined as
```
std::pair<size_t, std::vector<tools::wallet2::transfer_details>> wallet2::export_outputs() const
```
and writing
```
auto output = exports_outputs()
```
is much cleaner/easy to read. Compiler will take care of efficiency and return value optimization. There will be no excess copy.

Reply:
Actually, `auto outputs = exports_outputs()` will invoke move constructors for std::pair and std::vector here, so it's guaranteed that there will be no excess copy. It's better to remove &

Reply:
@xiphon did you read this last bit of the conversation?

Reply:
@SChernykh @NanoAkron  I doubt there is any difference in the ASM. On most ABIs, the callee chooses between move, copy or RVO with the caller providing the address. `const auto`, `auto` and `const auto&` only differ in how the caller can use the value after calling the function. `const auto&` has the advantage that if the return signature changes to `const &`, it still does the most optimal thing. I usually don't do this myself, but it can be good practice in some situations (particularly templated code).
 
However, if this value (`output`) were returned in this function (`export_outputs_to_str()`), then `auto` is better than `const auto` or `const auto&` because if the compiler cannot perform RVO it is required to copy instead of move. I've been making copy-constructors deleted and using `foo clone() const` (stolen from Rust!) instead as it catches these types of hidden copies (but isn't always possible with STL or boost code).

> P.S. I would rather rewrite the last line of export_outputs() to return std::make_pair(offset, std::move(outs)); if you really want to have no excess copies here.

Yup.